File: engines/py/anza/engine/file/load.py
from dataclasses import dataclass
from pathlib import Path
from typing import Any, Union
from ...crypto import digest
from ..slot import Chunk, extract, string
from .read import text


@dataclass
class Template:
    name: str
    path: Path
    raw: str
    digest: str
    chunks: list[Chunk]

    def bind(self, params: Any = None) -> str:
        return string(self.chunks, params)

    def render(self, params: Any = None) -> str:
        return string(self.chunks, params)


def one(root_dir: Union[str, Path], rel_path: Union[str, Path]) -> Template:
    root = Path(root_dir)
    full_path = root / rel_path
    raw = text(full_path)
    hash_hex = digest.hex(raw)
    chunks = extract(raw)
    normalized_name = str(Path(rel_path)).replace("\\", "/").lstrip("/")

    return Template(
        name=normalized_name,
        path=full_path,
        raw=raw,
        digest=hash_hex,
        chunks=chunks,
    )


def all(root_dir: Union[str, Path]) -> dict[str, Template]:
    root = Path(root_dir)
    templates: dict[str, Template] = {}

    if not root.exists():
        return templates

    for file_path in root.rglob("*.html"):
        if file_path.is_file():
            rel_path = file_path.relative_to(root)
            tpl = one(root, rel_path)
            templates[tpl.name] = tpl

    return templates

File: engines/rust/src/engine/file/load.rs
use {
  super::read,
  crate::{
    crypto::blake,
    engine::slot::{bind::{self, Data, Params}, extract, Chunk},
    errors::Result,
  },
  serde::Serialize,
  std::{
    collections::HashMap,
    path::{Path, PathBuf},
  },
  walkdir::WalkDir,
};

#[derive(Debug, Clone)]
pub struct Template {
  pub name: String,
  pub path: PathBuf,
  pub raw: String,
  pub digest: [u8; 32],
  pub chunks: Vec<Chunk>,
}

impl Template {
  /// Bind any parameter source (Value, tuples, arrays, slices, HashMaps, or () for empty)
  pub fn bind<P: Params + ?Sized>(&self, params: &P) -> Result<String> {
    bind::string(&self.chunks, params)
  }

  /// Render directly from any struct implementing `serde::Serialize`
  pub fn render<T: Serialize>(&self, data: &T) -> Result<String> {
    bind::string(&self.chunks, &Data(data))
  }
}

pub fn one(root: &Path, rel_path: &Path) -> Result<Template> {
  let full_path = root.join(rel_path);
  let raw = read::text(&full_path)?;
  let digest = blake::digest::content(raw.as_bytes());
  let chunks = extract(&raw)?;

  let name = rel_path
    .to_string_lossy()
    .trim_start_matches('/')
    .to_string();

  Ok(Template {
    name,
    path: full_path,
    raw,
    digest,
    chunks,
  })
}

pub fn all(root: &Path) -> Result<HashMap<String, Template>> {
  let mut map = HashMap::new();
  if !root.exists() {
    return Ok(map);
  }

  for entry in WalkDir::new(root).into_iter().filter_map(|e| e.ok()) {
    let path = entry.path();
    if path.is_file() && path.extension().map_or(false, |ext| ext == "html") {
      if let Ok(rel) = path.strip_prefix(root) {
        let tpl = one(root, rel)?;
        map.insert(tpl.name.clone(), tpl);
      }
    }
  }

  Ok(map)
}
